### PR TITLE
Migrated conditional dependencies to PEP 508 format

### PR DESCRIPTION
**Pull request by agronholm**
This uses a much more recent (and better) format for conditional dependencies than the previous PR.
I had to add ez_setup.py because setup.py now requires the latest setuptools to work.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -30,16 +30,14 @@
...
 
-import sys
 import re
 import os
 import platform
-from setuptools import setup, find_packages
 
-version = getattr(sys, "version_info", (0,))
+import ez_setup
+ez_setup.use_setuptools('20.3')
```
Comment (meejah) [inline setup.py]: Is it possible to just do "setup_requires=['setuptools>=20.3']" or so instead of including ez_setup...?

**Comment by agronholm** [comment]
Sigh...looks like pip hasn't vendored a recent enough `packaging` version yet :(

File: setup.py (hunk 1 above)
Comment (agronholm) [inline setup.py]: No. If you do that, the invocation of setup() will fail due to the requirements parse error before setuptools has a chance to update itself.

**Comment by agronholm** [comment]
Scratch that, 8.1.0 should have a good version vendored. Investigating further.

**Comment by agronholm** [comment]
This smells like a number of bugs in pip.

**Comment by oberstet** [comment]
Seems like Travis has pip 6.0.7 =(

https://travis-ci.org/crossbario/crossbar/jobs/116565855#L139

**Comment by agronholm** [comment]
I don't like it but I have to admit that this PR should probably be deferred until after the supporting tools are ready.

**Comment by oberstet** [comment]
@agronholm thanks for the CAA! got it.

>  this PR should probably be deferred until after the supporting tools are ready.

Agreed.

Link: the PR addresses https://github.com/crossbario/crossbar/issues/525

**Comment by agronholm** [comment]
This PR should fix the pip bug I'm seeing: https://github.com/pypa/pip/pull/3307

**Comment by oberstet** [comment]
@agronholm Travis is failing probably because of pip too old. What if you add a `pip install -U pip` here https://github.com/crossbario/crossbar/blob/master/.travis.yml#L5 ?

**Comment by agronholm** [comment]
@oberstet it won't help because even the latest pip will fail when doing `pip install .[crossbar]` because parts of pip still don't support PEP 508 requirements. The PR I linked to should fix that. And even when that happens, upgrading pip manually in the Travis build will only fix the Travis build. People with older pip versions will still have trouble installing the wheels. 

My (previous PR)[https://github.com/crossbario/crossbar/pull/677] _almost_ works, but the wheel installation fails on Travis because the name of the `platform_python_implementation` environment marker used to be called `python_implementation` but in packaging 16.5 (used by setuptools >= 20.2) it was renamed to its proper name. So if I use the new spec compliant name, the wheels will install using recent versions of pip but will introduce breakage for users of older pips (like the one used by Travis). On the other hand, if I use the older name, it will work only with older versions of pip/setuptools. So no matter what I do I lose :(

**Comment by oberstet** [comment]
Oh, well. I see.

I guess we all agree that the PEP508 approach is better than what we do now in setup.py, and that we do want to have wheels.

But we can't have it now because of the issues you explained.

Mmh. So what do we do?

**Comment by agronholm** [comment]
I'll try asking this on distutils-sig and see if they come up with something.

**Comment by hawkowl** [comment]
nooooo don't use ez_setup :(

dstufft has mentioned that if we do it conventionally in the setup.py, and overload it in setup.cfg, then the wheels will operate correctly on older pips, at the cost of duplicating dependencies.

**Comment by agronholm** [comment]
Can you point me to documentation on how to accomplish this? And how does this circumvent the problem with `platform_python_implementation` vs `python_implementation`?

**Comment by hawkowl** [comment]
I am not sure; let me try something out.

**Comment by agronholm** [comment]
According to my research, setuptools 18.5 is the minimum setuptools version that supports the `platform_python_implementation` marker. This is the minimum version needed to do either sdist or bdist_wheel with these markers present. However, pip versions older than that should support them in wheel metadata. So what is the oldest version of pip that crossbar needs to support?

It also seems that PEP 508 support is very incomplete in tooling -- the wheels created from this branch are missing the conditionals entirely. My other PR won't pass checks on Travis because Travis has setuptools older than 18.5. The use of ez_setup would fix that. @hawkowl 's PR is, in my opinion, a hack that should be avoided as the extra maintenance burden is not worth the very modest speed boost gained on installation speed, at least in case of a pure Python project like Crossbar.

**Comment by oberstet** [comment]
Using ez_setup ... nooo.

**Comment by agronholm** [comment]
Can you elaborate on why using ez_setup is bad? It's not like it replaces your current setuptools installation. It just conditionally downloads a newer one to the current directory and imports setuptools from that.

**Comment by oberstet** [comment]
ez_setup is bad, because Python packaging/setup stuff is already way too fucked up, and mixing in yet another thing (ez_setup) IMO will make things worth.

**Comment by agronholm** [comment]
The only purpose of ez_setup is to ensure that a new enough setuptools is used to run setup.py. It's transparent to the user and causes no action to be taken if the already installed setuptools is recent enough.

**Comment by agronholm** [comment]
I'd like to add that most users will never even hit ez_setup because they'll be installing the wheel.

**Comment by oberstet** [comment]
@agronholm I know how ez_setup works, and I don't like it;) A user should have pip installed, and a recent one, and the setup of a package (like Crossbar.io) should NOT fiddle with my setuptools/pip installation while installing the pacakge.

I think it's not worth for now fiddling with old pip and what not just to make a Crossbar.io wheel, which won't cut it anyway: on Linux, there are no binary wheels at all, and CB nowerdays has multiple hard dependencies on binary extensions (NaCl, LMDB, ..).

In other words: for now, you must have a C toolchain to install Crossbar.io from pip. You don't need a toolchain when you install Crossbar.io from our binary packages (deb/rpm, ..). And the latter is the recommended way to install it.

What was your original, concrete itch again?

**Comment by agronholm** [comment]
Pardon me, but from your reply it seems that my intentions were completely misunderstood. The reason for using ez_setup is to ensure the availability of a recent enough setuptools version for the PEP 508 requirement specifiers to be correctly processed when crossbar is installed from source (and not from wheel). For example, if you install directly from git, you would need a recent enough setuptools for the install to succeed if PEP 508 `install_requires` are to be supported. If ez_setup is not used, the user is expected to understand how to deal with the cryptic error message and upgrade their setuptools installation. If install failing on older setuptools versions is acceptable then ez_setup is not needed.

What are the minimum setuptools and pip versions expected from users of Crossbar anyway?

**Comment by oberstet** [comment]
@agronholm Let's just say the minimum versions are the current ones:

```console
(python279_1)oberstet@thinkpad-t430s:~/scm/crossbario/crossbar$ python
Python 2.7.9 (default, Jan 18 2015, 21:40:29) 
[GCC 4.8.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import pip
>>> pip.__version__
'8.1.1'
>>> import setuptools
>>> setuptools.__version__
'20.6.7'
```

I will check if I can make Travis update setuptools/pip before installing CB. If so, I am gonna remove the `OLD_SETUPTOOLS` stuff in `setup.py` and make above versions the officially supported minimal ones.

**Comment by oberstet** [comment]
@agronholm Another note: soonish, we will **pin** all Crossbar.io dependencies to exact point releases - and that will include setuptools.

Crossbar.io is supposed to be installed either into a virtualenv of its own - or installed via binary packages which bundle everything including Python and the kitchen sink anyway.

**Comment by oberstet** [comment]
@agronholm Another note: soonish, we will **pin** all Crossbar.io dependencies to exact point releases - and that will include setuptools.

Crossbar.io is supposed to be installed either into a virtualenv of its own - or installed via binary packages which bundle everything including Python and the kitchen sink anyway.

**Comment by oberstet** [comment]
see here https://github.com/crossbario/crossbar/issues/543 - we will pin all direct as well as indirect dependencies

**Comment by agronholm** [comment]
Just so you're aware: pinning setuptools won't help with the PEP 508 requirements problem because setuptools parses the install requirements *before* it tries to process the setup dependencies. Stupid design, yes, but that's the reality we live in.

Pinning dependencies exactly in Crossbar will cause me (and possibly other users) a *major* headache. My Asphalt framework has a testing dependency on Crossbar and if it suddenly starts pinning dependencies to exact versions, I will be unable to upgrade mine.

**Comment by oberstet** [comment]
Yes, it won't help for setuptools itself. But I figured that our plans for pinning would be of interest to you in this context. But having Crossbar.io as a Python level dependency is wrong. Cb should be treated as a blackbox, much like Nginx. Imagine I would find time and reasons some day, and I'd rewrite CB in C++ or Rust. There is no way you can have a Python level dep to a Rust written thing ..

Could you describe your Asphalt framework, and how/why it depends on CB at the Python level?

**Comment by agronholm** [comment]
Because I do integration testing with it in the asphalt-wamp subproject. See this function here: https://github.com/asphalt-framework/asphalt-wamp/blob/master/asphalt/wamp/utils.py#L40
I launch an ad-hoc Crossbar instance and run all integration tests through it. You're right in that if it wasn't written in Python it'd be much more of a PITA to do this.

**Comment by meejah** [comment]
@agronholm For my integration testing things I'm doing something very similar (although with py.test, and fixtures) and simply have the "crossbar" fixture rely on a "virtualenv" fixture, so that the "crossbar instance under test" is in its own virtualenv.

For sure slightly more pain than just a spawn in the same ("parent"?) virtualenv, but also nice separation and does actually treat the "crossbar" process more as a black-box (i.e. the same fixture would work if it was calling a compiled program; just the 'install' part becomes something besides 'make a new virtualenv'). This has an additional advantage: the 'parent' testing framework can (and does) use Autobahn libraries, but it can still test e.g. a released version of Autobahn while using a "master" or branch version itself...

**Comment by agronholm** [comment]
Ok @meejah, any chance you could share those fixtures?

**Comment by oberstet** [comment]
@agronholm This functional/integration test infrastructure for Crossbar.io itself is not open-source.

**Comment by meejah** [comment]
@agronholm Those particular fixtures are in a private repository :( but they're pretty similar to the code you pointed to -- just instead of ``sys.executable`` it's ``/path/to/venv/bin/crossbar``

**Comment by oberstet** [comment]
I have a PR that seems to work on Travis too: https://github.com/crossbario/crossbar/pull/707

Closing this one ..
